grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQuery.java

```diff
@@ -299,45 +297,39 @@ public void unify(Map<VarName, VarName> unifiers) {
         rule.unify(atom);
         ReasonerQueryImpl ruleBody = rule.getBody();
         ReasonerAtomicQuery ruleHead = rule.getHead();
+        Set<VarName> varsToRetain = rule.hasDisconnectedHead()? ruleBody.getVarNames() : ruleHead.getVarNames();
 
         subGoals.add(this);
-        Stream<Map<VarName, Concept>> subs = ruleBody.computeJoin(
-                subGoals,
-                cache,
-                dCache,
-                materialise,
-                differentialJoin);
-
-        Set<NotEquals> filters = ruleBody.getFilters();
-        Stream<Map<VarName, Concept>> answers = subs
-                .filter(a -> nonEqualsFilter(a, filters));
+        Stream<Map<VarName, Concept>> answers = ruleBody
+                .computeJoin(subGoals, cache, dCache, materialise, differentialJoin)
+                .flatMap(a -> varFilterFunction.apply(a, varsToRetain))
+                .distinct();
 
         if (materialise || rule.requiresMaterialisation()) {
-            Set<VarName> varsToRetain = rule.hasDisconnectedHead()? ruleBody.getVarNames() : ruleHead.getVarNames();
             LazyIterator<Map<VarName, Concept>> known = ruleHead.lazyLookup(cache);
             LazyIterator<Map<VarName, Concept>> dknown = ruleHead.lazyLookup(dCache);
             Stream<Map<VarName, Concept>> newAnswers = answers
-                    .flatMap(a -> varFilterFunction.apply(a, varsToRetain))
-                    .distinct()
                     .filter(a -> knownFilter(a, known.stream()))
                     .filter(a -> knownFilter(a, dknown.stream()))
                     .flatMap(ruleHead::materialise);
+            answers = dCache.record(ruleHead, newAnswers);
+        }
 
+        //if query not exactly equal to the rule head, do some conversion
+        if (!this.equals(ruleHead)) {
+            Set<VarName> vars = this.getVarNames();
+            Set<Map<VarName, VarName>> permutationUnifiers = getPermutationUnifiers(ruleHead.getAtom());
+            Set<IdPredicate> unmappedIdPredicates = atom.getUnmappedIdPredicates();
             Set<TypeAtom> mappedTypeConstraints = atom.getMappedTypeConstraints();
-            answers = dCache.record(ruleHead, newAnswers)
-                    .filter(a -> entityTypeFilter(a, mappedTypeConstraints));
+            Set<TypeAtom> unmappedTypeConstraints = atom.getUnmappedTypeConstraints();
+            answers = this.getIdPredicateAnswerStream(answers)
```

Comment:
With big chunks of code like this a one, a comment like your reply can clarify a lot. Not every line needs a comment but in cases like this I think it would be warranted.